rust/main/agents/validator/src/server/merkle_tree_insertions/list_merkle_tree_insertions.rs

```diff
@@ -0,0 +1,229 @@
+use axum::{
+    extract::{Query, State},
+    http::StatusCode,
+    routing::get,
+    Router,
+};
+use derive_new::new;
+use serde::{Deserialize, Serialize};
+
+use hyperlane_base::db::{HyperlaneDb, HyperlaneRocksDB};
+
+use crate::server::utils::{ServerErrorResponse, ServerResult, ServerSuccessResponse};
+
+#[derive(Clone, Debug, new)]
+pub struct ServerState {
+    pub db: HyperlaneRocksDB,
+}
+
+impl ServerState {
+    pub fn router(self) -> Router {
+        Router::new()
+            .route("/merkle_tree_insertions", get(handler))
+            .with_state(self)
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct QueryParams {
+    pub leaf_index_start: u32,
+    pub leaf_index_end: u32,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct TreeInsertion {
+    pub leaf_index: u32,
+    pub message_id: String,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ResponseBody {
+    pub merkle_tree_insertions: Vec<TreeInsertion>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ResponseErrorBody {
+    pub message: String,
+}
+
+/// Fetch merkle tree insertion into the database
+pub async fn handler(
+    State(state): State<ServerState>,
+    Query(query_params): Query<QueryParams>,
+) -> ServerResult<ServerSuccessResponse<ResponseBody>, ResponseErrorBody> {
+    let QueryParams {
```

Comment:
hopefully most of this logic can be deduplicated with the one in the relayer endpoint

Reply:
Updated!